### ISSUE TITLE
no-this-before-super should disallow declaring an arrow function that uses this before super()

### ISSUE THREAD

**Issue by sltam** [eslint/eslint#9201]
**ESLint Version:**
4.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
[ESLint online demo](https://eslint.org/demo/#eyJ0ZXh0IjoibmV3IGNsYXNzIGV4dGVuZHMgY2xhc3Mge30ge1xuXHRjb25zdHJ1Y3RvcigpIHtcblx0XHQoKCkgPT4gdGhpcykoKVxuXHRcdHN1cGVyKClcblx0fVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Nywic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnNvbGUiOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZW1wdHkiOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)

**What did you do? Please include the actual source code causing the issue.**

```js
new class extends class {} {
	constructor() {
		(() => this)()
		super()
	}
}
```

**What did you expect to happen?**
Warn me I am using `this` before `super`, even if I do not call the arrow function.

**What actually happened? Please include the actual, raw output from ESLint.**
`Lint-free!`

Try running it in a browser, you get `ReferenceError: Must call super constructor in derived class before accessing 'this' or returning from derived constructor`

**Comment by platinumazure** [comment]
This does seem like a bug to me. Function expressions and declarations might not be bound to the same `this` but arrow functions are bound to the lexical `this`. Hoping for a second opinion in case I'm missing something. Thanks for reporting!

**Comment by not-an-aardvark** [comment]
I think this is happening because we're using code path analysis to detect when `this` is referenced before `super` when executing the code, rather than just checking whether `this` appears at an earlier location in the file. As a result, we can't do analysis across function boundaries.

In general, it's not possible to statically determine whether `this` will actually be referenced before `super` in cases like this, because we don't know when the function containing `this` will actually be called. For example:

```js
class Foo extends Bar {
    constructor() {
        doSomething(() => this);
        super();
    }
}
```

The example above will only be a problem if `doSomething` executes its argument immediately, which we can't know when linting the file.

I think the code path analysis approach is a better choice than the "does `super` appear at an earlier location in the file than `this`" approach, because it will also be able to catch errors like:

```js
class Foo extends Bar {
    constructor() {
        if (condition) {
            super();
        } else {
            this;
        }
    }
}
```
